keps/sig-scheduling/3902-decoupled-taint-manager/README.md

```diff
@@ -0,0 +1,309 @@
+
+<p align="center"><image src="new-controllers.png" width=60%></p>
+
+### Custom taint-manager
+
+Opting out from the default `TaintManger`, Kubernetes operators will be able to implement custom logic to handle `NoExecute` taint such as delegating the authority for the termination for some classes of workloads providing high-availability protection guarantees. They can also scale out the usage of `NoExcute` to more use-cases than `NotReady` nodes (e.g. unhealthy network, etc.).
+
+<p align="center"><image src="custom-taint-manager.png" width=90%></p>
+
+### Implementation
+
+A new `NodeLifeCycleManager` is implemented by removing taint-manager related code from `controller/nodelifecycle/node_lifecycle_controller.go`.
+
+A new `NoExecuteTaintManager` is created a first-class controller managed by` kube-controller-manager`. It’s implementation is based on the current taint-manager  `controller/nodelifecycle/taint-manager.go`.  A new flag is added 
+
+`// NoExecuteTaintManagerConfiguration contains elements describing NoExecuteTaintManagerConfiguration`
+`type NoExecuteTaintManagerConfiguration struct {`
+`// enables the default NoExecuteTaintManager. MUST be synced with the`
+`// corresponding flag of the kube-apiserver.`
+`EnableNoExecuteTaintManager bool`
+`  ...`
+`}`
+
+The creation and startup of the default taint-manager is performed by `kube-controller-manager` . 
+
+`func startNoExecuteTaintManager(ctx context.Context, controllerContext ControllerContext) (controller.Interface, bool, error) {`
+`    if !controllerContext.ComponentConfig.NoExecuteTaintManager.EnableNoExecuteTaintManager {
+        return nil, false, nil
+    }`
+
+```
+taintManager, err = taintmanager.NewNoExecuteTaintManager(ctx, kubeClient, nc.podLister, nc.nodeLister, nc.getPodsAssignedToNode)
+        nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+            AddFunc: controllerutil.CreateAddNodeHandler(func(node *v1.Node) error {
+                nc.taintManager.NodeUpdated(nil, node)
+                return nil
+            }),
+            UpdateFunc: controllerutil.CreateUpdateNodeHandler(func(oldNode, newNode *v1.Node) error {
+                nc.taintManager.NodeUpdated(oldNode, newNode)
+                return nil
+            }),
+            DeleteFunc: controllerutil.CreateDeleteNodeHandler(func(node *v1.Node) error {
+                nc.taintManager.NodeUpdated(node, nil)
+                return nil
+            }),
+        })
+    }
+        if err != nil {
+        return nil, true, fmt.Errorf("failed to start the taint manager: %v", err)
+    }
+  }  
+```
+
+### Test Plan
+
+* Unit and integration tests
+    * Add nodes taints
+    * Act on `NoExecute` to delete running pods
+    * Respect taint tolerations 
+* Verify passing the existing E2E and conformance tests.
+    * Taint a node 
+    * Evict the correct set of running pods
```

Comment:
You may need to detail the tests section. There is a updated version requiring documenting the current code coverage, etc. You can refer to other latest KEP for references.

Reply:
How about 

* Unit tests
    * Enable `TaintManager`
    * Disable `TaintManager`
    * Add `NoExecute` taints
    * Act on `NoExecute` to delete running pods
    * Respect taint tolerations
* Integration tests
    * Enable `TaintManager`, add `NoExecute` taints to evict the correct set of running pods on nodes
    * Disable `TaintManager`, add `NoExecute` has no impact on running pods
* E2E tests
    * Verify passing the existing E2E and conformance tests.